lib/spack/spack/builder.py

```diff
@@ -0,0 +1,512 @@
+    """Class decorator used to register the default builder
+    for a given build-system.
+
+    Args:
+        build_system_name (str): name of the build-system
+    """
+
+    def _decorator(cls):
+        cls.build_system = build_system_name
+        BUILDER_CLS[build_system_name] = cls
+        return cls
+
+    return _decorator
+
+
+def create(pkg):
+    """Given a package object with an associated concrete spec,
+    return the builder object that can install it.
+
+    Args:
+         pkg (spack.package_base.PackageBase): package for which we want the builder
+    """
+    if id(pkg) not in _BUILDERS:
+        _BUILDERS[id(pkg)] = _create(pkg)
+    return _BUILDERS[id(pkg)]
+
+
+class _PhaseAdapter(object):
+    def __init__(self, builder, phase_fn):
+        self.builder = builder
+        self.phase_fn = phase_fn
+
+    def __call__(self, spec, prefix):
+        return self.phase_fn(self.builder.pkg, spec, prefix)
+
+
+def _create(pkg):
+    package_module = inspect.getmodule(pkg)
+    package_buildsystem = buildsystem_name(pkg)
+    default_builder_cls = BUILDER_CLS[package_buildsystem]
+    builder_cls_name = default_builder_cls.__name__
+    builder_cls = getattr(package_module, builder_cls_name, None)
+    if builder_cls:
+        return builder_cls(pkg)
+
+    # Specialized version of a given buildsystem can subclass some
+    # base classes and specialize certain phases or methods or attributes.
+    # In that case they can store their builder class as a class level attribute.
+    # See e.g. AspellDictPackage as an example.
+    base_cls = getattr(pkg, builder_cls_name, default_builder_cls)
+
+    # From here on we define classes to construct a special builder that adapts to the
+    # old, single class, package format. The adapter forwards any call or access to an
+    # attribute related to the installation procedure to a package object wrapped in
+    # a class that falls-back on calling the base builder if no override is found on the
+    # package. The semantic should be the same as the method in the base builder were still
+    # present in the base class of the package.
+
+    class _ForwardToBaseBuilder(object):
+        def __init__(self, wrapped_pkg_object, root_builder):
+            self.wrapped_package_object = wrapped_pkg_object
+            self.root_builder = root_builder
+
+            package_cls = type(wrapped_pkg_object)
+            wrapper_cls = type(self)
+            bases = (package_cls, wrapper_cls)
+            new_cls_name = package_cls.__name__ + "Wrapper"
+            new_cls = type(new_cls_name, bases, {})
+            new_cls.__module__ = package_cls.__module__
+            self.__class__ = new_cls
+            self.__dict__.update(wrapped_pkg_object.__dict__)
+
+        def __getattr__(self, item):
+            default_item = getattr(base_cls, item)
+            setattr(type(self.root_builder), item, default_item)
+            result = getattr(self.root_builder, item)
```

Comment:
If we have an "old-style" Package like

```
class Foo(CMakePackage):
    def cmake_args(...)
        ...
    ...

# No builder defined alongside
```

Then this will "absorb" `Foo`, so if this object (e.g. `pkg_with_dispatcher) calls `cmake_args`, we will end up here (via `forward_method_to_getattr`). It looks like what this is doing is:

* Grabbing `cmake_args` from the builder (on line 103 - that's what `base_cls` is here)
* Setting `cmake_args` on the Builder Adapter to be that
* Using getattr on that

A couple points:

* It seems like `Foo.cmake_args` would not be used here (this retrieves from the builder rather than ever examining `self`)
* Why use `setattr` on `root_builder` every time? I'd assume you would not have to use `setattr` twice if you call Foo.cmake_args` twice

Reply:
> Why use setattr on root_builder every time? I'd assume you would not have to use setattr twice if you call Foo.cmake_args twice

That's not needed. I'll push a change using:
```python
result = getattr(super(type(self.root_builder), self.root_builder), item)
if item in super(type(self.root_builder), self.root_builder).phases:
    result = _PhaseAdapter(self.root_builder, result)
```

Reply:
> Then this will "absorb" Foo, so if this object (e.g. pkg_with_dispatcher) calls cmake_args, we will end up here (via forward_method_to_getattr`).

I think there's a fundamental misunderstanding here. This class is added at the end of the class hierarchy, so:
```python
Spack version 0.19.0.dev0
Python 3.8.10, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec("openjpeg").concretized()
>>> s.package.cmake_args()
['-DBUILD_CODEC:BOOL=OFF', '-DBUILD_MJ2:BOOL=OFF', '-DBUILD_THIRDPARTY:BOOL=OFF']
>>> builder = s.package.builder
>>> builder.cmake_args()
['-DBUILD_CODEC:BOOL=OFF', '-DBUILD_MJ2:BOOL=OFF', '-DBUILD_THIRDPARTY:BOOL=OFF']
```

Reply:
Should be done